fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexBuildState.java

```diff
@@ -0,0 +1,160 @@
+ * <ul>
+ *     <li> Records scanned does not include the records scanned in the {@link OnlineIndexer} whose {@link OnlineIndexer.Builder#setTrackProgress(boolean)} is set to false. </li>
+ *     <li> Records added or deleted during the index build progress are not included in records scanned but are in records in total. </li>
+ *     <li> Records in total uses the number of all records in the store. In some cases, the real number of records that need to be scanned is less than this.</li>
+ * </ul>
+ * @see #getIndexBuildStateAsync(FDBRecordStore, Index)
+ */
+@API(API.Status.UNSTABLE)
+public class IndexBuildState {
+    @Nonnull
+    private final IndexState indexState;
+    @Nullable
+    private final Long recordsScanned;
+    @Nullable
+    private final Long recordsInTotal;
+
+    /**
+     * Get the build progress ({@link IndexBuildState}) of the given index in the given record store asynchronously.
+     * @param store the record store containing the index
+     * @param index the index needed to be checked
+     * @return a future that completes to the index build state
+     */
+    public static CompletableFuture<IndexBuildState> getIndexBuildStateAsync(FDBRecordStore store, Index index) {
+        IndexState indexState = store.getIndexState(index);
+        if (indexState != IndexState.WRITE_ONLY) {
+            return CompletableFuture.completedFuture(new IndexBuildState(indexState));
+        }
+        return getRecordsScannedAsync(store, index).thenCompose(scannedRecords -> {
+            try {
+                return store.getSnapshotRecordCount()
+                        .thenApply(totalRecords -> new IndexBuildState(indexState, scannedRecords, totalRecords));
+            } catch (Exception ex) {
+                // getSnapshotRecordCount failed, very likely it is because there is no suitable COUNT type index
+                // defined.
+                return CompletableFuture.completedFuture(new IndexBuildState(indexState, scannedRecords, null));
+            }
+        });
+    }
+
+    /**
+     * Get the number of records successfully scanned and processed during the online index build process
+     * asynchronously.
+     * <p>
+     * This does not include the records scanned in the {@link OnlineIndexer} whose
+     * {@link OnlineIndexer.Builder#setTrackProgress(boolean)} is set to false.
+     * </p>
+     * @param store the record store containing the index
+     * @param index the index needed to be checked
+     * @return a future that completes to the total records scanned.
+     */
+    public static CompletableFuture<Long> getRecordsScannedAsync(FDBRecordStore store, Index index) {
+        return store.getContext().ensureActive()
+                .get(OnlineIndexer.indexBuildScannedRecordsSubspace(store, index).getKey())
+                .thenApply(FDBRecordStore::decodeRecordCount);
+    }
+
+    /**
+     * Get the index state.
+     * @return the index state
+     */
+    public IndexState getIndexState() {
```

Comment:
These accessor methods are missing nullability annotations on their return types (especially important for the nullable ones).